File: python/auth_helper.py
"""
auth_helper — Token cache để tái sử dụng token giữa các lần chạy
==================================================================
Thay vì gọi authenticate() mỗi lần chạy script, module này:
  1. Load token đã lưu từ file token_cache.json (nếu có)
  2. Nếu chưa có → authenticate lần đầu và lưu xuống file
  3. Nếu token hết hạn → refresh và lưu lại
  4. Nếu token còn hạn → dùng trực tiếp, không gọi API

Cách dùng (sync):
    from auth_helper import ensure_auth
    with Auth(config) as auth:
        ensure_auth(auth)

Cách dùng (async):
    from auth_helper import ensure_auth_async
    async with AsyncAuth(config) as auth:
        await ensure_auth_async(auth)
"""

import json
import os

from ssi_sdk.models import Token

TOKEN_FILE = os.path.join(os.path.dirname(__file__), "token_cache.json")


def load_token() -> Token | None:
    """Load token từ file, trả về None nếu file không tồn tại."""
    if not os.path.exists(TOKEN_FILE):
        return None
    with open(TOKEN_FILE, "r", encoding="utf-8") as f:
        _token = json.load(f)
    return Token.from_dict(_token)


def save_token(token: Token) -> None:
    """Lưu token xuống file."""
    with open(TOKEN_FILE, "w", encoding="utf-8") as f:
        json.dump(token.to_dict(), f, indent=2)
    print(f"Token đã lưu vào {TOKEN_FILE}")


def ensure_auth(auth, otp: str | None = None) -> None:
    """Đảm bảo auth có token hợp lệ (sync)."""
    cached_token = load_token()
    if cached_token is not None:
        auth.token_manager.set_token(cached_token)
        if not auth.is_token_expired:
            print("Token còn hạn, dùng token từ file.")
            return
        print("Token đã hết hạn, đang refresh...")
        try:
            cached_token = auth.refresh()
            save_token(cached_token)
            print("Refresh token thành công.")
            return
        except Exception as e:
            print(f"Refresh token thất bại ({e}), tiến hành xác thực lại...")

    print("Không tìm thấy token hợp lệ, đang thực hiện quy trình xác thực & OTP...")
    if otp:
        cached_token = auth.authenticate(otp=otp)
    else:
        print("=== Yêu cầu OTP (Request OTP) ===")
        otp_res = auth.request_otp()
        transaction_id = None
        if isinstance(otp_res, dict):
            data_map = otp_res.get("data", {})
            if isinstance(data_map, dict):
                transaction_id = data_map.get("transactionId") or data_map.get("transaction_id")
            if not transaction_id:
                transaction_id = otp_res.get("transactionId") or otp_res.get("transaction_id")

        if transaction_id:
            print(f"[Smart OTP] Transaction ID: {transaction_id}")
            print("Vui lòng mở app SSI iBoard trên điện thoại và bấm APPROVE (Xác nhận)...")
            print("SDK đang Polling chờ bạn bấm phê duyệt...")
            auth.ensure_authenticated(transaction_id=transaction_id, poll_interval=5, poll_max_retries=6)
        else:
            user_otp = input("Vui lòng nhập mã OTP 6 số: ").strip()
            auth.authenticate(otp=user_otp)

    if auth.token:
        save_token(auth.token)
        print("Authenticate thành công, token đã lưu.")


async def ensure_auth_async(auth, otp: str | None = None) -> None:
    """Đảm bảo auth có token hợp lệ (async)."""
    cached_token = load_token()
    if cached_token is not None:
        await auth.token_manager.set_token(cached_token)
        if not auth.is_token_expired:
            print("Token còn hạn, dùng token từ file.")
            return
        print("Token đã hết hạn, đang refresh...")
        try:
            cached_token = await auth.refresh()
            save_token(cached_token)
            print("Refresh token thành công.")
            return
        except Exception as e:
            print(f"Refresh token thất bại ({e}), tiến hành xác thực lại...")

    print("Không tìm thấy token hợp lệ, đang thực hiện quy trình xác thực & OTP...")
    if otp:
        cached_token = await auth.authenticate(otp=otp)
    else:
        print("=== Yêu cầu OTP (Request OTP) ===")
        otp_res = await auth.request_otp()
        transaction_id = None
        if isinstance(otp_res, dict):
            data_map = otp_res.get("data", {})
            if isinstance(data_map, dict):
                transaction_id = data_map.get("transactionId") or data_map.get("transaction_id")
            if not transaction_id:
                transaction_id = otp_res.get("transactionId") or otp_res.get("transaction_id")

        if transaction_id:
            print(f"[Smart OTP] Transaction ID: {transaction_id}")
            print("Vui lòng mở app SSI iBoard trên điện thoại và bấm APPROVE (Xác nhận)...")
            print("SDK đang Polling chờ bạn bấm phê duyệt...")
            await auth.ensure_authenticated(transaction_id=transaction_id, poll_interval=5, poll_max_retries=6)
        else:
            user_otp = input("Vui lòng nhập mã OTP 6 số: ").strip()
            await auth.authenticate(otp=user_otp)

    if auth.token:
        save_token(auth.token)
        print("Authenticate thành công, token đã lưu.")

File: node/auth_helper.js
/**
 * auth_helper — Token cache de tai su dung token giua cac lan chay
 * ==================================================================
 * Thay vi goi authenticate() moi lan chay script, module nay:
 *   1. Load token da luu tu file token_cache.json (neu co)
 *   2. Neu chua co -> authenticate lan dau (co the can OTP) va luu xuong file
 *   3. Neu token het han -> refresh va luu lai
 *   4. Neu token con han -> dung truc tiep, khong goi API
 *
 * Cach dung:
 *   import { Auth } from '@ssi.developer/ssi-sdk';
 *   import { config } from './config.js';
 *   import { ensureAuth } from './auth_helper.js';
 *
 *   const auth = new Auth(config);
 *   await ensureAuth(auth);          // market data — khong can OTP
 *   await ensureAuth(auth, OTP);     // trading/streaming — can OTP lan dau
 */

import fs from 'node:fs';
import path from 'node:path';
import readline from 'node:readline';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SHARED_TOKEN_FILE = path.join(__dirname, '..', 'shared_token.json');
const TOKEN_FILE = path.join(__dirname, 'token_cache.json');

export function loadToken() {
  for (const file of [SHARED_TOKEN_FILE, TOKEN_FILE]) {
    if (!fs.existsSync(file)) continue;
    try {
      return JSON.parse(fs.readFileSync(file, 'utf-8'));
    } catch {
      continue;
    }
  }
  return null;
}

export function saveToken(token) {
  fs.writeFileSync(TOKEN_FILE, JSON.stringify(token, null, 2), 'utf-8');
  console.log(`Token da luu vao ${TOKEN_FILE}`);
}

function askQuestion(query) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => {
    rl.question(query, (ans) => {
      rl.close();
      resolve(ans);
    });
  });
}

/**
 * Dam bao `auth` co token hop le.
 * @param {import('@ssi.developer/ssi-sdk').Auth} auth
 * @param {string} [otp] OTP cho lan authenticate dau tien.
 */
export async function ensureAuth(auth, otp) {
  let token = loadToken();

  if (token != null) {
    auth.setToken(token);
    if (auth.tokenManager.isAuthenticated()) {
      console.log('Token con han, dung token tu file.');
      return;
    }

    console.log('Token da het han, dang refresh...');
    try {
      token = await auth.refresh();
      saveToken(token);
      console.log('Refresh token thanh cong.');
      return;
    } catch (err) {
      console.log(`Refresh token that bai (${err.message}), tien hanh xac thuc lai...`);
    }
  }

  console.log('Khong tim thay token hop le, dang thuc hien quy trinh xac thuc & OTP...');
  if (otp) {
    token = await auth.authenticate(otp);
  } else {
    console.log('=== Yeu cau OTP (Request OTP) ===');
    const otpRes = await auth.requestOtp();
    let transactionId = otpRes?.data?.transactionId || otpRes?.transactionId;

    if (transactionId) {
      console.log(`[Smart OTP] Transaction ID: ${transactionId}`);
      console.log('Vui long mo ung dung SSI tren dien thoai va bam APPROVE (Xac nhan)...');
      console.log('SDK dang Polling cho ban bam phe duyet...');
      await auth.ensureAuthenticated(undefined, transactionId, 5000, 6);
      token = auth.getToken();
    } else {
      const userOtp = await askQuestion('Vui long nhap ma OTP 6 so: ');
      if (userOtp.trim()) {
        token = await auth.authenticate(userOtp.trim());
      }
    }
  }

  if (auth.getToken()) {
    saveToken(auth.getToken());
    console.log('Authenticate thanh cong, token da luu.');
  }
}

